backend/pikpak/pikpak.go

```diff
@@ -0,0 +1,1749 @@
+	if err != nil {
+		return err
+	}
+
+	if srcParentID != dstParentID {
+		// Do the move
+		err = f.moveObjects(ctx, []string{srcID}, dstParentID)
+		if err != nil {
+			return fmt.Errorf("couldn't dir move: %w", err)
+		}
+	}
+
+	// Can't copy and change name in one step so we have to check if we have
+	// the correct name after copy
+	if srcLeaf != dstLeaf {
+		_, err = f.renameObject(ctx, srcID, dstLeaf)
+		if err != nil {
+			return fmt.Errorf("dirmove: couldn't rename moved dir: %w", err)
+		}
+	}
+	srcFs.dirCache.FlushDir(srcRemote)
+	return nil
+}
+
+// Creates from the parameters passed in a half finished Object which
+// must have setMetaData called on it
+//
+// Returns the object, leaf, dirID and error.
+//
+// Used to create new objects
+func (f *Fs) createObject(ctx context.Context, remote string, modTime time.Time, size int64) (o *Object, leaf string, dirID string, err error) {
+	// Create the directory for the object if it doesn't exist
+	leaf, dirID, err = f.dirCache.FindPath(ctx, remote, true)
+	if err != nil {
+		return
+	}
+	// Temporary Object under construction
+	o = &Object{
+		fs:      f,
+		remote:  remote,
+		size:    size,
+		modTime: modTime,
+	}
+	return o, leaf, dirID, nil
+}
+
+// Move src to this remote using server-side move operations.
+//
+// This is stored with the remote path given.
+//
+// It returns the destination Object and a possible error.
+//
+// Will only be called if src.Fs().Name() == f.Name()
+//
+// If it isn't possible then return fs.ErrorCantMove
+func (f *Fs) Move(ctx context.Context, src fs.Object, remote string) (fs.Object, error) {
+	srcObj, ok := src.(*Object)
+	if !ok {
+		fs.Debugf(src, "Can't move - not same remote type")
+		return nil, fs.ErrorCantMove
+	}
+	err := srcObj.readMetaData(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	srcLeaf, srcParentID, err := srcObj.fs.dirCache.FindPath(ctx, srcObj.remote, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create temporary object
+	dstObj, dstLeaf, dstParentID, err := f.createObject(ctx, remote, srcObj.modTime, srcObj.size)
+	if err != nil {
+		return nil, err
+	}
+
+	if srcParentID != dstParentID {
+		// Do the move
+		if err = f.moveObjects(ctx, []string{srcObj.id}, dstParentID); err != nil {
+			return nil, err
+		}
+	}
+	dstObj.id = srcObj.id
+
+	var info *api.File
+	if srcLeaf != dstLeaf {
+		// Rename
+		info, err = f.renameObject(ctx, srcObj.id, dstLeaf)
+		if err != nil {
+			return nil, fmt.Errorf("move: couldn't rename moved file: %w", err)
+		}
+	} else {
+		// Update info
+		info, err = f.getFile(ctx, dstObj.id)
+		if err != nil {
+			return nil, fmt.Errorf("move: couldn't update moved file: %w", err)
+		}
+	}
+	return dstObj, dstObj.setMetaData(info)
+}
+
+// copy objects
+func (f *Fs) copyObjects(ctx context.Context, IDs []string, dirID string) (err error) {
+	if len(IDs) == 0 {
+		return nil
+	}
+	req := api.RequestBatch{
+		Ids: IDs,
+		To:  &map[string]string{"parent_id": parentIDForRequest(dirID)},
```

Comment:
We don't usually use pointers to `map` as they are a reference type and can be `nil`?